nni/algorithms/hpo/gridsearch_tuner.py

```diff
@@ -2,209 +2,241 @@
 # Licensed under the MIT license.
 
 """
-gridsearch_tuner.py including:
-    class GridSearchTuner
+Grid search tuner for hyper-parameter optimization.
+
+For categorical parameters this tuner fully explore all combinations.
+For numerical parameters it samples them at progressively decreased intervals.
+
+Use this tuner if you have abundant resource and want to find strictly optimal parameters.
+
+Grid search tuner has no argument.
 """
 
-import copy
+__all__ = ['GridSearchTuner']
+
 import logging
+import math
+
 import numpy as np
+from scipy.special import erfinv  # pylint: disable=no-name-in-module
 
 import nni
-from nni.common.hpo_utils import validate_search_space
+from nni.common.hpo_utils import ParameterSpec, deformat_parameters, format_search_space
 from nni.tuner import Tuner
-from nni.utils import convert_dict2tuple
 
-TYPE = '_type'
-CHOICE = 'choice'
-VALUE = '_value'
-
-logger = logging.getLogger('grid_search_AutoML')
+_logger = logging.getLogger('nni.tuner.gridsearch')
+
+##
+# Grid search is a simple algorithm if only categorical parameters are considered.
+# But to support continuous space, things get tricky.
+#
+# To support continuous space, we divide search process into "epochs".
+# The first epoch only explores middle point of uniform and normal parameters.
+# When first epoch is fully explored, the algorithm starts second epoch,
+# where it divides non-categorical spaces by adding quartile points into the grid.
+# Then in third epoch it adds [1/8, 3/8, 5/8, 7/8], and so on.
+#
+# We divide normal distributed spaces using inverse function of CDF.
+# For example the 1/4 point of a normal distribution is defined as X where `normal_cdf(X) = 1/4`.
+#
+# Here is an example:
+#
+#   search space:
+#     x: choices(5, 7)
+#     y: normal(0, 1)
+#     z: quniform(2, 3, 1)
```

Comment:
quniform is discrete space, you treat it a continuous space?

Reply:
The example has already make it clear.
`3/4 is eliminated due to duplication`